pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -137,23 +137,42 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, pus
 	envoyFilterPatches := push.EnvoyFilters(proxy)
 	cb := NewClusterBuilder(proxy, updates, configgen.Cache)
 	instances := proxy.ServiceInstances
+
 	for key := range updates.ConfigsUpdated {
 		// get service that changed
-		service := push.ServiceForHostname(proxy, host.Name(key.Name))
-		// SidecarScope.Service will return nil if the proxy doesn't care about the service OR it was deleted.
-		// we can cross reference with WatchedResources to figure out which services were deleted.
-		if service == nil {
-			// WatchedResources.ResourceNames will contain the names of the clusters it is subscribed to. We can
-			// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>
-			// so, we can check if the cluster names contains the service name, and determine if it is deleted by that
-			for _, n := range watched.ResourceNames {
-				_, _, svcHost, _ := model.ParseSubsetKey(n)
-				if svcHost == host.Name(key.Name) {
-					removedClusterNames = append(removedClusterNames, n)
-				}
+		if key.Kind == gvk.Service {
+			service := push.ServiceForHostname(proxy, host.Name(key.Name))
+			// SidecarScope.ServiceForHostname will return nil if the proxy doesn't care about the service OR it was deleted.
+			// we can cross reference with WatchedResources to figure out which services were deleted.
+			if service == nil {
+				// we assume a service was deleted, it doesn't hurt to have positives for removedResources when a resource included
+				// does not exist.
+				// WatchedResources.ResourceNames will contain the names of the clusters it is subscribed to. We can
+				// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>
+				// so, we can check if the cluster names contains the service name, and determine if it is deleted by that
+				removedClusterNames = append(removedClusterNames, getRemovedClusterName(watched.ResourceNames, key.Name))
+			} else {
+				services = append(services, service)
 			}
 		} else {
-			services = append(services, service)
+			// destination rules
+			cfg := proxy.SidecarScope.DestinationRuleByName(key.Name)
+			if cfg == nil {
+				// a destinationrule was deleted, get the populated destinationRule from PrevSidecarScope
+				prevCfg := proxy.PrevSidecarScope.DestinationRuleByName(key.Name)
+				if prevCfg == nil {
+					// something went wrong, skip -- we shouldn't be nil here
+					break
+				}
+				dr := prevCfg.Spec.(*networking.DestinationRule)
+				removedClusterNames = append(removedClusterNames, getRemovedClusterName(watched.ResourceNames, dr.Host))
+			} else {
+				// destination exists, was updated
+				// generate cluster based on service associated with destinationrule
+				dr := cfg.Spec.(*networking.DestinationRule)
+				service := push.ServiceForHostname(proxy, host.Name(dr.Host))
+				services = append(services, service)
```

Comment:
I don't think this accounts for duplicates - either a Service and DR for the same service change, or multiple DRs for the same service changing

Reply:
yeah worst case scenario we generate duplicate clusters I think, but will fix.

Reply:
wait, is there actual harm in this? seems like we detect duplicates with `pilot_duplicate_envoy_clusters`